### PR TITLE
Add methods to get the index of a triple when searching

### PR DESCRIPTION
**Pull request by ate47**
I've merged the [triplePosition](https://github.com/rdfhdt/hdt-java/tree/triplePosition) branch with the 3.0.0 to be able to use the new methods of this branch. 

## API changes

The usage of the API is the same, but now we have two more methods,

In ID/String iterators, to get the position of the last triple returned by next()

```java
long getLastTriplePosition();
```

In Triple interface, to get the opposite operation:

```java
TripleID findTriple(long position);
```

## Position

The position is the index on the last line of the tree,

Capture d’écran 2022-02-21 à 10 55 25

For example, if we are ``searching(2, *, *)``, the result iterator / position would be:

```
searching(2, *, *)=
  (2, 2, 1), index=1
  (2, 3, 3), index=2
  (2, 3, 4), index=3
  (2, 5, 2), index=4
```

Same computation for ``searching(*, *, 1)``

```
searching(*, *, 1)=
  (2, 2, 1), index=1
  (3, 1, 1), index=5
  (4, 1, 1), index=6
```

## Memory and computation power usage

Because to get the position of a triple in the iterators ZFOQ and Y requires to do a binary search, the computation of the position is done when the call of the method is made, for Buffered (with HDT get with HDTManager.mapHDT instead of HDTManager.loadHDT) and SequentialSearch iterators, the value to compute the position are stored to also avoid the computation.

So if we don't use the methods, the memory consumed is 1-3 longs for non buffered iterator and buffer-size * (1-3 longs) for buffered iterators and the nearly 0 computation are made except to store the position computation arguments.

## Tests

10 tests were made times the number of dictionary, so 10*4=40 tests, to test these features:

- 2 tests to check with and without buffering to check if we have a difference with or without buffer.
- 8 tests to check the 2^3 SPO patterns (???, S??, ... SP?, SPO), with buffer.
 

## Difference with #141

Compared to Ali's PR (#141), we don't have to use a different search method to use the ids, so we don't have a ``getIndex()`` that can't be used if we aren't in a wrong search. But we have to store the computation arguments of the position when we are using the index.

### CODE REVIEW

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/iterator/DictionaryTranslateIterator.java [hunk 1]
```diff
@@ -154,4 +154,9 @@ public ResultEstimationType numResultEstimation() {
 		return iterator.numResultEstimation();
 	}
 
+	@Override
+	public long getLastTriplePosition() {
+		return iterator.getLastTriplePosition();
+	}
+
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/iterator/DictionaryTranslateIterator.java:161]: why did this change? I do not see a relation with the new feature ....

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriples.java [hunk 2]
```diff
@@ -701,6 +702,31 @@ public String getType() {
 		return HDTVocabulary.TRIPLES_TYPE_BITMAP;
 	}
 
+	@Override
+	public TripleID findTriple(long position) {
+		if (position == 0) {
+			// remove this special case so we can use position-1
+			return new TripleID(1, 1, 1);
+		}
```
Comment (D063520) [inline hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriples.java:710]: what is the aim of this?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/iterator/DictionaryTranslateIterator.java (hunk 1 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/iterator/DictionaryTranslateIterator.java:161]: I think you were talking about the [TestArray.java](https://github.com/rdfhdt/hdt-java/pull/150/commits/f7da5495075afe83e89602bdbea92103905dbe13#diff-c11e4dba31593ee7bd70eb0fc3713d617cc6e3d2e21c7f8f63597eef1d3e1255) and [MapArray.java](https://github.com/rdfhdt/hdt-java/pull/150/commits/f7da5495075afe83e89602bdbea92103905dbe13#diff-a708f6236dd75bed4d84dddfbe99257e86bad2944e8087a19727c3cb45d65701) files, they were in a previous commit of the old branch, I've removed them

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriples.java (hunk 2 above)
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/triples/impl/BitmapTriples.java:713]: A little forget of my side, it's fixed and I've added a test to check this

File: hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorPositionTest.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+
+@RunWith(Parameterized.class)
+public class BitmapTriplesIteratorPositionTest {
+
+    private static final Field ITERATOR_SUB;
+    private static final Field ITERATOR_SUB_BUFFER;
+    private static final Field ITERATOR_SUB_SEQ;
+
```
Comment (D063520) [inline hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorPositionTest.java:34]: what is this?

File: hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTSortedUtils.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
+package org.rdfhdt.hdt.triples.impl.utils;
```
Comment (D063520) [inline hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTSortedUtils.java:1]: why did this class change?

File: hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTTestUtils.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
+package org.rdfhdt.hdt.triples.impl.utils;
```
Comment (D063520) [inline hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTTestUtils.java:1]: same here ... why is this class changed in the pull request?

File: hdt-java-core/src/main/java/org/rdfhdt/hdt/iterator/DictionaryTranslateIteratorBuffer.java [hunk 6]
```diff
@@ -47,31 +46,40 @@
  * 
  */
 public class DictionaryTranslateIteratorBuffer implements IteratorTripleString {
+	private static class IndexedTripleID {
```
[the remark this answers is not in the document]
Comment (ate47) [reply hdt-java-core/src/main/java/org/rdfhdt/hdt/iterator/DictionaryTranslateIteratorBuffer.java]: it was a quick name to merge the words index and tripleID, I've just renamed it in https://github.com/rdfhdt/hdt-java/pull/150/commits/55b2d8fd786a8432f910bc72647c0ed7f200adc8

File: hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorPositionTest.java (hunk 3 above)
Comment (ate47) [reply hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorPositionTest.java:34]: It's the iterator fields of the Wrapper DictionaryTranslateIterator, DictionaryTranslateIteratorBuffer, SequentialSearchIteratorTripleID, it's used in the [BitmapTriplesIteratorPositionTest#printIterator(Object)](https://github.com/ate47/hdt-java/blob/tripleposition-next/hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/BitmapTriplesIteratorPositionTest.java#L79) method to print an iterator and all the sub iterators we know, you can read the [test results](https://github.com/ate47/hdt-java/runs/5286529509?check_suite_focus=true#step:5:1505), on the line 1505:

```
Search with pattern:S?O
[0] Used iterator: class org.rdfhdt.hdt.iterator.DictionaryTranslateIteratorBuffer
[1] Used iterator: class org.rdfhdt.hdt.iterator.SequentialSearchIteratorTripleID
[2] Used iterator: class org.rdfhdt.hdt.triples.impl.BitmapTriplesIterator
```

File: hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTSortedUtils.java (hunk 4 above)
Comment (ate47) [reply hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTSortedUtils.java:1]: I've added this new HDT utility class to create a fixed HDT with easy to get positions.

File: hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTTestUtils.java (hunk 5 above)
Comment (ate47) [reply hdt-java-core/src/test/java/org/rdfhdt/hdt/triples/impl/utils/HDTTestUtils.java:1]: Same here, but the position and the HDT are gotten from a list.

**Pull request merged**